File: src/python/finished/smallest_missing_non_negative_integer_after_operations.py
from typing import List

# https://leetcode.com/problems/smallest-missing-non-negative-integer-after-operations/


class Solution:
    def findSmallestInteger(self, nums: List[int], value: int) -> int:
        classes = [0] * value

        for n in nums:
            clas = n % value
            classes[clas] += 1

        min_cnt = min(classes)
        for k, cnt in enumerate(classes):
            if cnt == min_cnt:
                return cnt * value + k
        return -1

File: src/rust/finished/smallest_missing_non_negative_integer_after_operations.rs
impl Solution {
  pub fn find_smallest_integer(nums: Vec<i32>, value: i32) -> i32 {
      nums.iter()
          .fold(vec![0; value as usize], |mut cs, n| {
              let class = n.rem_euclid(value);
              cs[class as usize] += 1;
              cs
          })
          .iter()
          .enumerate()
          .fold(i32::MAX, |mex, (class, &size)| {
              mex.min(size * value + class as i32)
          })
  }
}
